File: code/packages/python/ruby-parser/tests/test_ruby_parser.py
"""Tests for the Ruby Parser.

These tests verify that the grammar-driven parser, when loaded with the
``ruby.grammar`` file, correctly parses Ruby source code into ASTs.

The key insight being tested is that **no new parser code was written**.
The same ``GrammarParser`` that handles Python handles Ruby — only the
grammar file changed. The parser produces generic ``ASTNode`` trees that
reflect the grammar's rule structure.

Understanding the AST Structure
-------------------------------

The grammar-driven parser produces ``ASTNode`` objects where:

- ``rule_name`` is the name of the grammar rule that matched
  (e.g., ``"program"``, ``"assignment"``, ``"expression"``)
- ``children`` is a list of ``ASTNode`` and ``Token`` objects that
  were matched by the rule's body

For example, parsing ``x = 1 + 2`` produces::

    ASTNode(rule_name="program", children=[
        ASTNode(rule_name="statement", children=[
            ASTNode(rule_name="assignment", children=[
                Token(NAME, "x"),
                Token(EQUALS, "="),
                ASTNode(rule_name="expression", children=[
                    ASTNode(rule_name="term", children=[
                        ASTNode(rule_name="factor", children=[
                            Token(NUMBER, "1")
                        ])
                    ]),
                    Token(PLUS, "+"),
                    ASTNode(rule_name="term", children=[
                        ASTNode(rule_name="factor", children=[
                            Token(NUMBER, "2")
                        ])
                    ])
                ])
            ])
        ])
    ])

Test Organization
-----------------

1. **Assignments** — ``x = 1``, ``x = 1 + 2``
2. **Operator precedence** — ``1 + 2 * 3`` vs ``(1 + 2) * 3``
3. **Multiple statements** — programs with more than one statement
4. **Method calls** — ``puts("hello")``, ``add(1, 2)``
5. **Factory function** — ``create_ruby_parser()``
"""

from __future__ import annotations

from lang_parser import ASTNode
from lexer import Token, TokenType

from ruby_parser import create_ruby_parser, parse_ruby


# ============================================================================
# Helpers — make AST inspection more readable
# ============================================================================


def find_nodes(node: ASTNode, rule_name: str) -> list[ASTNode]:
    """Recursively find all descendant nodes with the given rule name.

    This is a depth-first search through the AST. It is useful for
    finding specific constructs (like all "factor" nodes) without
    having to manually walk the tree.

    Args:
        node: The root node to search from.
        rule_name: The grammar rule name to search for.

    Returns:
        A list of all matching ASTNode objects, in depth-first order.
    """
    results: list[ASTNode] = []
    if node.rule_name == rule_name:
        results.append(node)
    for child in node.children:
        if isinstance(child, ASTNode):
            results.extend(find_nodes(child, rule_name))
    return results


def find_tokens(node: ASTNode) -> list[Token]:
    """Recursively collect all Token leaves from an AST.

    This "flattens" the tree back into a token list, which is useful
    for verifying that all tokens from the source are accounted for
    in the AST.

    Args:
        node: The root node to collect tokens from.

    Returns:
        A list of all Token objects in the tree, in depth-first order.
    """
    tokens: list[Token] = []
    for child in node.children:
        if isinstance(child, Token):
            tokens.append(child)
        elif isinstance(child, ASTNode):
            tokens.extend(find_tokens(child))
    return tokens


# ============================================================================
# Test: Simple Assignments
# ============================================================================


class TestAssignments:
    """Test parsing of Ruby assignment statements.

    Assignments are one of the most fundamental constructs in any
    language: ``variable = expression``. The parser must recognize the
    NAME, EQUALS, and expression parts, and build an AST that reflects
    the assignment structure.
    """

    def test_simple_assignment(self) -> None:
        """Parse ``x = 1`` — the simplest possible assignment.

        Expected AST structure:
            program -> statement -> assignment -> [NAME, EQUALS, expression]
        """
        ast = parse_ruby("x = 1")
        assert ast.rule_name == "program"

        # The program should have one statement child
        statements = find_nodes(ast, "statement")
        assert len(statements) == 1

        # That statement should contain an assignment
        assignments = find_nodes(ast, "assignment")
        assert len(assignments) == 1

        # The assignment should contain a NAME token with value "x"
        assignment_tokens = find_tokens(assignments[0])
        names = [t for t in assignment_tokens if t.type == TokenType.NAME]
        assert len(names) == 1
        assert names[0].value == "x"

    def test_assignment_with_arithmetic(self) -> None:
        """Parse ``x = 1 + 2`` — assignment with a binary expression.

        The right-hand side ``1 + 2`` should be parsed as an expression
        containing two terms connected by a PLUS operator.
        """
        ast = parse_ruby("x = 1 + 2")
        assert ast.rule_name == "program"

        # Find the expression inside the assignment
        expressions = find_nodes(ast, "expression")
        assert len(expressions) >= 1

        # The expression should contain a PLUS token
        expr_tokens = find_tokens(expressions[0])
        plus_tokens = [t for t in expr_tokens if t.type == TokenType.PLUS]
        assert len(plus_tokens) == 1

    def test_assignment_with_string(self) -> None:
        """Parse ``name = "ruby"`` — assignment with a string value."""
        ast = parse_ruby('name = "ruby"')

        assignments = find_nodes(ast, "assignment")
        assert len(assignments) == 1

        tokens = find_tokens(assignments[0])
        string_tokens = [t for t in tokens if t.type == TokenType.STRING]
        assert len(string_tokens) == 1
        assert string_tokens[0].value == "ruby"

    def test_assignment_with_variable(self) -> None:
        """Parse ``y = x`` — assignment from another variable."""
        ast = parse_ruby("y = x")

        assignments = find_nodes(ast, "assignment")
        assert len(assignments) == 1

        tokens = find_tokens(assignments[0])
        names = [t for t in tokens if t.type == TokenType.NAME]
        # Two NAMEs: "y" (target) and "x" (value)
        assert len(names) == 2
        assert names[0].value == "y"
        assert names[1].value == "x"


# ============================================================================
# Test: Operator Precedence
# ============================================================================


class TestOperatorPrecedence:
    """Test that the parser correctly handles operator precedence.

    The grammar encodes precedence through rule nesting:
    - ``expression`` handles ``+`` and ``-`` (lowest precedence)
    - ``term`` handles ``*`` and ``/`` (higher precedence)
    - ``factor`` handles atoms and parentheses (highest precedence)

    This means ``1 + 2 * 3`` parses as ``1 + (2 * 3)`` because ``*``
    is handled at the ``term`` level, which is deeper in the parse tree.
    """

    def test_addition(self) -> None:
        """Parse ``1 + 2`` — simple addition."""
        ast = parse_ruby("1 + 2")

        # Should have an expression with two terms and a PLUS
        expressions = find_nodes(ast, "expression")
        assert len(expressions) >= 1

        tokens = find_tokens(expressions[0])
        values = [t.value for t in tokens]
        assert "1" in values
        assert "+" in values
        assert "2" in values

    def test_multiplication_before_addition(self) -> None:
        """Parse ``1 + 2 * 3`` — multiplication has higher precedence.

        In the AST, ``2 * 3`` should be grouped together inside a ``term``
        node, while ``1`` and ``2 * 3`` are children of the ``expression``.
        """
        ast = parse_ruby("1 + 2 * 3")

        # The top-level expression should have a PLUS
        expressions = find_nodes(ast, "expression")
        expr_direct_tokens = [
            c for c in expressions[0].children if isinstance(c, Token)
        ]
        plus_tokens = [t for t in expr_direct_tokens if t.value == "+"]
        assert len(plus_tokens) == 1

        # There should be term nodes containing the multiplication
        terms = find_nodes(ast, "term")
        # Find the term that contains the STAR
        star_terms = [
            t for t in terms
            if any(
                isinstance(c, Token) and c.value == "*"
                for c in t.children
            )
        ]
        assert len(star_terms) == 1

    def test_parentheses_override_precedence(self) -> None:
        """Parse ``(1 + 2) * 3`` — parentheses override precedence.

        The parentheses force ``1 + 2`` to be computed first. In the AST,
        the addition should appear inside a ``factor`` (which contains
        a parenthesized expression), and the multiplication should be
        at the ``term`` level.
        """
        ast = parse_ruby("(1 + 2) * 3")

        # The outer structure should be a term with STAR
        terms = find_nodes(ast, "term")
        star_terms = [
            t for t in terms
            if any(
                isinstance(c, Token) and c.value == "*"
                for c in t.children
            )
        ]
        assert len(star_terms) == 1

        # Inside should be a factor containing LPAREN, expression, RPAREN
        factors = find_nodes(ast, "factor")
        paren_factors = [
            f for f in factors
            if any(
                isinstance(c, Token) and c.type == TokenType.LPAREN
                for c in f.children
            )
        ]
        assert len(paren_factors) == 1

    def test_subtraction_and_division(self) -> None:
        """Parse ``10 - 4 / 2`` — division has higher precedence than subtraction."""
        ast = parse_ruby("10 - 4 / 2")

        all_tokens = find_tokens(ast)
        values = [t.value for t in all_tokens]
        assert "10" in values
        assert "-" in values
        assert "4" in values
        assert "/" in values
        assert "2" in values

    def test_chained_addition(self) -> None:
        """Parse ``1 + 2 + 3`` — left-associative addition."""
        ast = parse_ruby("1 + 2 + 3")

        expressions = find_nodes(ast, "expression")
        assert len(expressions) >= 1

        # Should have two PLUS tokens in the expression
        expr_tokens = [
            c for c in expressions[0].children
            if isinstance(c, Token) and c.value == "+"
        ]
        assert len(expr_tokens) == 2


# ============================================================================
# Test: Multiple Statements
# ============================================================================


class TestMultipleStatements:
    """Test parsing of programs with multiple statements.

    The ``program`` rule matches ``{ statement }``, meaning zero or more
    statements. The parser must correctly separate statements (typically
    by newlines) and parse each one independently.
    """

    def test_two_assignments(self) -> None:
        """Parse two assignment statements separated by a newline."""
        ast = parse_ruby("x = 1\ny = 2")
        assert ast.rule_name == "program"

        assignments = find_nodes(ast, "assignment")
        assert len(assignments) == 2

    def test_three_statements(self) -> None:
        """Parse three statements of mixed types."""
        ast = parse_ruby("x = 1\ny = x + 2\nz = y * 3")

        assignments = find_nodes(ast, "assignment")
        assert len(assignments) == 3

    def test_expression_statement(self) -> None:
        """Parse a bare expression as a statement (no assignment)."""
        ast = parse_ruby("1 + 2")

        # Should have one expression_stmt
        expr_stmts = find_nodes(ast, "expression_stmt")
        assert len(expr_stmts) == 1

    def test_mixed_assignments_and_expressions(self) -> None:
        """Parse a mix of assignments and expression statements."""
        ast = parse_ruby("x = 5\nx + 1")

        statements = find_nodes(ast, "statement")
        assert len(statements) == 2


# ============================================================================
# Test: Method Calls
# ============================================================================


class TestMethodCalls:
    """Test parsing of Ruby method calls.

    Method calls in our Ruby grammar have the form:
        (NAME | KEYWORD) LPAREN [ expression { COMMA expression } ] RPAREN

    This handles both regular method calls like ``add(1, 2)`` and
    keyword-based calls like ``puts("hello")``.
    """

    def test_puts_with_string(self) -> None:
        """Parse ``puts("hello")`` — the quintessential Ruby output call."""
        ast = parse_ruby('puts("hello")')

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

        tokens = find_tokens(method_calls[0])
        # Should contain: KEYWORD(puts), LPAREN, STRING(hello), RPAREN
        keyword_tokens = [t for t in tokens if t.type == TokenType.KEYWORD]
        assert len(keyword_tokens) == 1
        assert keyword_tokens[0].value == "puts"

        string_tokens = [t for t in tokens if t.type == TokenType.STRING]
        assert len(string_tokens) == 1
        assert string_tokens[0].value == "hello"

    def test_method_call_with_name(self) -> None:
        """Parse ``add(1, 2)`` — a regular method call with two arguments."""
        ast = parse_ruby("add(1, 2)")

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

        tokens = find_tokens(method_calls[0])
        name_tokens = [t for t in tokens if t.type == TokenType.NAME]
        assert name_tokens[0].value == "add"

        number_tokens = [t for t in tokens if t.type == TokenType.NUMBER]
        assert len(number_tokens) == 2
        assert number_tokens[0].value == "1"
        assert number_tokens[1].value == "2"

    def test_method_call_no_args(self) -> None:
        """Parse ``hello()`` — a method call with no arguments."""
        ast = parse_ruby("hello()")

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

    def test_method_call_with_expression_arg(self) -> None:
        """Parse ``print(1 + 2)`` — a method call with an expression argument."""
        ast = parse_ruby("print(1 + 2)")

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

        # The argument should be parsed as an expression containing PLUS
        expressions = find_nodes(method_calls[0], "expression")
        assert len(expressions) >= 1

    def test_puts_with_variable(self) -> None:
        """Parse ``puts(x)`` — puts with a variable argument."""
        ast = parse_ruby("puts(x)")

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

        tokens = find_tokens(method_calls[0])
        names = [t for t in tokens if t.type == TokenType.NAME]
        assert any(t.value == "x" for t in names)


# ============================================================================
# Test: Factory Function
# ============================================================================


class TestCreateRubyParser:
    """Test the ``create_ruby_parser()`` factory function.

    While ``parse_ruby()`` is the simpler interface, ``create_ruby_parser()``
    gives access to the ``GrammarParser`` object, which is useful for
    advanced inspection or integration with custom pipelines.
    """

    def test_creates_parser(self) -> None:
        """The factory should return a GrammarParser with a parse method."""
        parser = create_ruby_parser("x = 1")
        assert hasattr(parser, "parse")

    def test_factory_produces_same_result(self) -> None:
        """The factory should produce the same AST as parse_ruby()."""
        source = "x = 1 + 2"
        ast_direct = parse_ruby(source)
        ast_factory = create_ruby_parser(source).parse()

        # Both should have the same rule_name and same number of children
        assert ast_direct.rule_name == ast_factory.rule_name
        assert len(ast_direct.children) == len(ast_factory.children)

    def test_factory_with_method_call(self) -> None:
        """Verify the factory works with method calls."""
        parser = create_ruby_parser('puts("world")')
        ast = parser.parse()
        assert ast.rule_name == "program"

        method_calls = find_nodes(ast, "method_call")
        assert len(method_calls) == 1

File: code/packages/lua/ruby_parser/tests/test_ruby_parser.lua
-- Tests for ruby_parser
-- ======================
--
-- Comprehensive busted test suite for the Ruby parser package.
--
-- Test coverage:
--   - Module loads and exposes the public API
--   - x = 5 → program → statement → assignment
--   - puts("hello") → method_call
--   - if x > 0: ... end → (via expression parse)
--   - Grammar first rule is "program"
--   - create_parser returns a GrammarParser
--   - get_grammar returns a grammar with rules
--   - Empty program is valid
--   - Arithmetic precedence: 1 + 2 * 3

-- Resolve sibling packages from the monorepo so busted can find them
-- without requiring a global luarocks install.
package.path = (
    "../src/?.lua;"                                             ..
    "../src/?/init.lua;"                                        ..
    "../../grammar_tools/src/?.lua;"                            ..
    "../../grammar_tools/src/?/init.lua;"                       ..
    "../../lexer/src/?.lua;"                                    ..
    "../../lexer/src/?/init.lua;"                               ..
    "../../state_machine/src/?.lua;"                            ..
    "../../state_machine/src/?/init.lua;"                       ..
    "../../directed_graph/src/?.lua;"                           ..
    "../../directed_graph/src/?/init.lua;"                      ..
    "../../ruby_lexer/src/?.lua;"                               ..
    "../../ruby_lexer/src/?/init.lua;"                          ..
    "../../parser/src/?.lua;"                                   ..
    "../../parser/src/?/init.lua;"                              ..
    package.path
)

local ruby_parser = require("coding_adventures.ruby_parser")

-- =========================================================================
-- Helper utilities
-- =========================================================================

--- Recursively collect all rule_names from an AST, in pre-order.
-- @param node  ASTNode  The root node to walk.
-- @return table         Ordered list of rule_name strings encountered.
local function collect_rule_names(node)
    local out = {}
    local function walk(n)
        if type(n) ~= "table" then return end
        if n.rule_name then
            out[#out + 1] = n.rule_name
            if n.children then
                for _, child in ipairs(n.children) do
                    walk(child)
                end
            end
        end
    end
    walk(node)
    return out
end

--- Find the first ASTNode with the given rule_name (depth-first).
-- @param node      ASTNode  Root to search.
-- @param rule_name string   Rule name to find.
-- @return ASTNode|nil
local function find_node(node, rule_name)
    if type(node) ~= "table" then return nil end
    if node.rule_name == rule_name then return node end
    if node.children then
        for _, child in ipairs(node.children) do
            local found = find_node(child, rule_name)
            if found then return found end
        end
    end
    return nil
end

--- Count all nodes with the given rule_name (full traversal).
-- @param node      ASTNode
-- @param rule_name string
-- @return number
local function count_nodes(node, rule_name)
    if type(node) ~= "table" then return 0 end
    local n = (node.rule_name == rule_name) and 1 or 0
    if node.children then
        for _, child in ipairs(node.children) do
            n = n + count_nodes(child, rule_name)
        end
    end
    return n
end

-- =========================================================================
-- Module surface
-- =========================================================================

describe("ruby_parser module", function()
    it("loads successfully", function()
        assert.is_not_nil(ruby_parser)
    end)

    it("exposes VERSION as a string", function()
        assert.is_string(ruby_parser.VERSION)
        assert.matches("^%d+%.%d+%.%d+$", ruby_parser.VERSION)
    end)

    it("exposes parse as a function", function()
        assert.is_function(ruby_parser.parse)
    end)

    it("exposes create_parser as a function", function()
        assert.is_function(ruby_parser.create_parser)
    end)

    it("exposes get_grammar as a function", function()
        assert.is_function(ruby_parser.get_grammar)
    end)

    it("get_grammar returns an object with rules", function()
        local g = ruby_parser.get_grammar()
        assert.is_not_nil(g)
        assert.is_table(g.rules)
        -- ruby.grammar has: program, statement, assignment,
        -- method_call, expression_stmt, expression, term, factor
        assert.is_true(#g.rules >= 6)
    end)

    it("grammar first rule is program", function()
        local g = ruby_parser.get_grammar()
        assert.are.equal("program", g.rules[1].name)
    end)
end)

-- =========================================================================
-- Root node
-- =========================================================================

describe("root AST node", function()
    it("parse returns a non-nil value", function()
        local ast = ruby_parser.parse("x = 5")
        assert.is_not_nil(ast)
    end)

    it("root node has rule_name == 'program'", function()
        local ast = ruby_parser.parse("x = 5")
        assert.are.equal("program", ast.rule_name)
    end)

    it("root node has children table", function()
        local ast = ruby_parser.parse("x = 5")
        assert.is_table(ast.children)
    end)

    it("program contains a statement node", function()
        local ast = ruby_parser.parse("x = 5")
        local stmt = find_node(ast, "statement")
        assert.is_not_nil(stmt, "expected 'statement' node")
    end)

    it("parses empty program (no statements)", function()
        local ast = ruby_parser.parse("")
        assert.are.equal("program", ast.rule_name)
        assert.is_table(ast.children)
    end)
end)

-- =========================================================================
-- Assignments
-- =========================================================================

describe("assignments", function()
    it("parses x = 5", function()
        local ast = ruby_parser.parse("x = 5")
        assert.are.equal("program", ast.rule_name)
        local assign = find_node(ast, "assignment")
        assert.is_not_nil(assign, "expected 'assignment' node")
    end)

    it("parses string assignment", function()
        local ast = ruby_parser.parse('name = "Alice"')
        assert.are.equal("program", ast.rule_name)
        local assign = find_node(ast, "assignment")
        assert.is_not_nil(assign)
    end)

    it("assignment contains expression", function()
        local ast = ruby_parser.parse("x = 42")
        local expr = find_node(ast, "expression")
        assert.is_not_nil(expr, "expected 'expression' inside assignment")
    end)

    it("parses multiple assignments", function()
        local ast = ruby_parser.parse("x = 1\ny = 2")
        assert.are.equal("program", ast.rule_name)
        local count = count_nodes(ast, "assignment")
        assert.are.equal(2, count)
    end)
end)

-- =========================================================================
-- Method calls
-- =========================================================================

describe("method calls", function()
    it('parses puts("hello")', function()
        local ast = ruby_parser.parse('puts("hello")')
        assert.are.equal("program", ast.rule_name)
        local mc = find_node(ast, "method_call")
        assert.is_not_nil(mc, "expected 'method_call' node")
    end)

    it("parses method call with numeric argument", function()
        local ast = ruby_parser.parse("puts(42)")
        assert.are.equal("program", ast.rule_name)
        local mc = find_node(ast, "method_call")
        assert.is_not_nil(mc)
    end)

    it("parses method call with expression argument", function()
        local ast = ruby_parser.parse("puts(1 + 2)")
        assert.are.equal("program", ast.rule_name)
        local mc = find_node(ast, "method_call")
        assert.is_not_nil(mc)
        local expr = find_node(ast, "expression")
        assert.is_not_nil(expr)
    end)
end)

-- =========================================================================
-- Expression statements
-- =========================================================================

describe("expression statements", function()
    it("parses numeric literal as expression_stmt", function()
        local ast = ruby_parser.parse("42")
        assert.are.equal("program", ast.rule_name)
        local es = find_node(ast, "expression_stmt")
        assert.is_not_nil(es, "expected 'expression_stmt' node")
    end)

    it("parses name as expression_stmt", function()
        local ast = ruby_parser.parse("x")
        assert.are.equal("program", ast.rule_name)
        local es = find_node(ast, "expression_stmt")
        assert.is_not_nil(es)
    end)
end)

-- =========================================================================
-- Expression precedence
-- =========================================================================

describe("expression precedence", function()
    it("1 + 2 * 3 → has expression and term nodes", function()
        -- Grammar encodes precedence: expression handles +/-, term handles */.
        local ast = ruby_parser.parse("r = 1 + 2 * 3")
        assert.are.equal("program", ast.rule_name)
        local expr = find_node(ast, "expression")
        assert.is_not_nil(expr, "expected 'expression' node")
        local term = find_node(ast, "term")
        assert.is_not_nil(term, "expected 'term' node")
    end)

    it("arithmetic produces factor nodes", function()
        local ast = ruby_parser.parse("r = 5 + 3")
        assert.are.equal("program", ast.rule_name)
        local factor = find_node(ast, "factor")
        assert.is_not_nil(factor, "expected 'factor' node")
    end)

    it("parses parenthesized expression", function()
        local ast = ruby_parser.parse("r = (2 + 3) * 4")
        assert.are.equal("program", ast.rule_name)
        local expr = find_node(ast, "expression")
        assert.is_not_nil(expr)
    end)

    it("deeply nested arithmetic", function()
        local ast = ruby_parser.parse("r = 1 + 2 + 3 + 4")
        assert.are.equal("program", ast.rule_name)
        local count = count_nodes(ast, "term")
        assert.is_true(count >= 4, "expected at least 4 term nodes")
    end)

    it("subtraction is at expression level", function()
        local ast = ruby_parser.parse("r = 10 - 3")
        assert.are.equal("program", ast.rule_name)
        local expr = find_node(ast, "expression")
        assert.is_not_nil(expr)
    end)

    it("division is at term level", function()
        local ast = ruby_parser.parse("r = 10 / 2")
        assert.are.equal("program", ast.rule_name)
        local term = find_node(ast, "term")
        assert.is_not_nil(term)
    end)
end)

-- =========================================================================
-- Multiple statements
-- =========================================================================

describe("multiple statements", function()
    it("parses two assignments", function()
        local ast = ruby_parser.parse("a = 1\nb = 2")
        assert.are.equal("program", ast.rule_name)
        local count = count_nodes(ast, "assignment")
        assert.are.equal(2, count)
    end)

    it("parses assignment and expression_stmt", function()
        local ast = ruby_parser.parse("x = 5\n42")
        assert.are.equal("program", ast.rule_name)
        assert.are.equal(1, count_nodes(ast, "assignment"))
        assert.are.equal(1, count_nodes(ast, "expression_stmt"))
    end)

    it("parses assignment and method call", function()
        local ast = ruby_parser.parse('x = 5\nputs("hello")')
        assert.are.equal("program", ast.rule_name)
        assert.are.equal(1, count_nodes(ast, "assignment"))
        assert.are.equal(1, count_nodes(ast, "method_call"))
    end)
end)

-- =========================================================================
-- create_parser
-- =========================================================================

describe("create_parser", function()
    it("returns a non-nil parser object", function()
        local p = ruby_parser.create_parser("x = 1")
        assert.is_not_nil(p)
    end)

    it("returned parser has a parse method", function()
        local p = ruby_parser.create_parser("x = 1")
        assert.is_function(p.parse)
    end)

    it("parsing via create_parser returns same root rule_name as parse()", function()
        local src  = "x = 42"
        local ast1 = ruby_parser.parse(src)
        local p    = ruby_parser.create_parser(src)
        local ast2, err = p:parse()
        assert.is_nil(err)
        assert.are.equal(ast1.rule_name, ast2.rule_name)
    end)
end)

-- =========================================================================
-- Grammar inspection
-- =========================================================================

describe("grammar inspection", function()
    it("grammar has assignment rule", function()
        local g = ruby_parser.get_grammar()
        local found = false
        for _, rule in ipairs(g.rules) do
            if rule.name == "assignment" then found = true; break end
        end
        assert.is_true(found, "expected 'assignment' rule in grammar")
    end)

    it("grammar has method_call rule", function()
        local g = ruby_parser.get_grammar()
        local found = false
        for _, rule in ipairs(g.rules) do
            if rule.name == "method_call" then found = true; break end
        end
        assert.is_true(found, "expected 'method_call' rule in grammar")
    end)

    it("grammar has expression rule", function()
        local g = ruby_parser.get_grammar()
        local found = false
        for _, rule in ipairs(g.rules) do
            if rule.name == "expression" then found = true; break end
        end
        assert.is_true(found, "expected 'expression' rule in grammar")
    end)
end)

-- =========================================================================
-- Error handling
-- =========================================================================

describe("error handling", function()
    it("raises an error for completely invalid input", function()
        assert.has_error(function()
            ruby_parser.parse("@@@ GARBAGE @@@")
        end)
    end)
end)
